### ISSUE TITLE
Sampling from a PWM

### ISSUE THREAD

**Issue by laserson** [vanheeringen-lab/gimmemotifs#204]
Is there a recommended way to "realize" a sequence from a PWM using random sampling? I would imagine something like

```
rng = np.random.default_rng()
motif.sample_from_pwm(rng)
```

that would generate the associated random sequence.

Alternatively, is there a recommended snippet of code to accomplish which is *fast*?

**Comment by Maarten-vd-Sande** [comment]
I'm not entirely sure I understand what you mean, sth like this? Idk if it's fast, but I don't really see how this could be a bottleneck in an analysis

```
import random

n_seqs = 10_000
nucs = [random.choices("ACGT", weights=motif.pwm[i], k=n_seqs) for i in range(len(motif.pwm))]
seqs = ["".join(nuc) for nuc in map(list, zip(*nucs))]
```

**Comment by laserson** [comment]
We generate A LOT of synthetic sequences in which we embed motifs. The code takes minutes to run (when it should ideally take seconds), and is bottlenecked by the random sampling step.

I thought I submitted a separate issue, but another problem is that the rows (or cols?) of the PWM don't sum to one. I have to renormalize some of them before sampling. I imagine this is an issue with the conversion of the floating point values to text and then back. Perhaps a renormalization can be included in the initialization?

**Comment by laserson** [comment]
We want to generate millions of sequences, and we do it with *all* motifs from a given motif library.

I created #209 to track the normalization.

**Comment by laserson** [comment]
fwiw, here is code I use, but could probably be faster:

```python
def sample_from_motif_pwm(motif: Motif, rng: np.random.Generator) -> str:
    pwm = np.asarray(motif.pwm)
    # apparently some of the positions don't sum to 1
    pwm /= pwm.sum(axis=1).reshape(-1, 1)
    return "".join([rng.choice(_nucs, p=p) for p in pwm])
```

**Comment by Maarten-vd-Sande** [comment]
The problem is that you are calling np.random for each sequence separately. Getting random stuff is expensive. The solution I posted on friday generates many sequences at once, and in turn takes less than 100/th of the time for e.g. 10.000 sequences. Either doing it with numpy or stdlib has no significant impact on the timings here.

However making 1.000.000 sequences still takes almost 3 seconds, and since we have 1.800 motifs this will take 1.5 hour! Using multiprocessing seems to make this process approximately another 1/3 faster, so it should still take +/- half an hour to generate a million sequences from all 1800 motifs! I don't see any easy way to get it faster than that, so I hope this suits your needs :smile: 

```
import random
from multiprocessing import Pool

import numpy as np

from gimmemotifs.motif import Motif,read_motifs


# either stdlib
def sample_from_motif_pwm_stdlib(motif, n_seqs):
    nucs = [random.choices("ACGT", weights=motif.pwm[i], k=n_seqs) for i in range(len(motif.pwm))]
    seqs = ["".join(nuc) for nuc in zip(*nucs)]
    return seqs

# or with numpy (similar speed)
def sample_from_motif_pwm_np(motif, n_seqs):
    pwm = np.asarray(motif.pwm)
    # apparently some of the positions don't sum to 1
    pwm /= pwm.sum(axis=1).reshape(-1, 1)
    
    nucs = [np.random.choice(["A", "C", "G", "T"], p=pwm[i], size=n_seqs) for i in range(len(pwm))]
    seqs = ["".join(nuc) for nuc in zip(*nucs)]

    return seqs


# make a multiprocessing pool, more processes make it faster, diminishing returns)
motifs = read_motifs()
n_seqs = 10_000
pool = Pool(processes=4)

# now run it in parallel!
now = time.time()
results = pool.starmap(sample_from_motif_pwm_stdlib, [(motif, n_seqs) for motif in motifs])
print(time.time() - now)

now = time.time()
results = pool.starmap(sample_from_motif_pwm_stdlib, [(motif, n_seqs) for motif in motifs])
print(time.time() - now)
```

I did not dip my feet in seeding the functions, and I leave that for you. I can imagine some strange stuff happens with multiprocessing. Probably you want to set the seed inside the function (as you did before). Good luck!
